llama-swap: init at 156, nixos/llama-swap: init module

06kellyjac opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

Packages llama-swap for nixpkgs

UI also working
<img width="800" height="800" alt="image" src="https://github.com/user-attachments/assets/1e74b200-234d-435c-a4dc-9401472535b7" />

Adopted module from @podium868909 , tweaked docs a bit, and added hardening copied from llama-cpp
Added module test with opt-in SmolLM2 model usage with model switching.

Either set 

or you can test with this file:

```nix
{
  lib ? (import ./. { }).lib,
  pkgs ? import ./. {
    config.allowUnfreePredicate =
      pkg:
      builtins.elem (lib.getName pkg) [
        "smollm2-135m" # try commenting this out
      ];
  },
}:
pkgs.nixosTests.llama-swap
```

`nix-build ./thefileyoumade.nix` in the copy of `nixpkgs` with this change.

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [X] [NixOS tests] in [nixos/tests].
  - [X] [Package tests] at `passthru.tests`.
  - Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [X] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - Package update: when the change is major or breaking.
- NixOS Release Notes
  - [X] Module addition: when adding a new NixOS module.
  - Module update: when the change is significant.
- [X] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

podium868909 committed:
maintainers: add podium868909

06kellyjac committed:
llama-swap: init module

Co-authored-by: podium868909 <89096245@proton.me>

06kellyjac committed:
llama-swap: init at 156

Changelog: https://github.com/mostlygeek/llama-swap/releases/tag/v156

acid-bong commented on pkgs/by-name/ll/llama-swap/package.nix:
```diff
@@ -0,0 +1,104 @@
+      git rev-parse HEAD > $out/COMMIT
+      # '0000-00-00T00:00:00Z'
+      date -u -d "@$(git log -1 --pretty=%ct)" "+'%Y-%m-%dT%H:%M:%SZ'" > $out/SOURCE_DATE_EPOCH
+      find "$out" -name .git -print0 | xargs -0 rm -rf
```
```suggestion
      rm -rf .git
```
the PWD is already the root dir

since the result will be the same, hash shouldn't change

06kellyjac replied:
this is a common pattern.

```
λ rg "\-name \.git \-print0" --count-matches | awk -F: '{s+=$2} END {print s}'
45
```

https://github.com/NixOS/nixpkgs/commit/9d7932e349942db74ec2c3a7030085c77d7a3716#diff-2781420d92792cbc2471e058d776f10a032c5519459f07df791f395b9bde29f5R12-R22

It'll find any `.git` dirs across the whole output and is the exact command ran if you don't enable `leaveDotGit`.
Since the intention is to: grab extra info, then go back to `leaveDotGit = false;` behavior, I'd rather keep it

https://github.com/NixOS/nixpkgs/blob/67bb48c4c8e327417d6d5aa7e538244b209e852b/pkgs/build-support/fetchgit/nix-prefetch-git#L348

acid-bong replied:
fair enough, in case there are submodules. consider this convo resolved

xNefas commented:
Having this issue when trying to grab the package from an input of which the URL is the PR's original repository.

```
nix log /nix/store/8bmwhg1hhbv1if6yh1ifwh6l6rhlncnp-llama-swap-149-go-modules.drv
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/jg46sjjnb3g57hdd3ijv70s5sfb35qgy-source
source root is source
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Executing npmConfigHook
Configuring npm

ERROR: no dependencies were specified
Hint: set `npmDeps` if using these hooks individually. If this is happening with `buildNpmPackage`, please open an issue.
```

`npmDeps` is clearly set in the original derivation's definition, yet it fails to build on my system. Refer to [my configuration](https://gitlab.com/thomasyou03/re-nixos/-/tree/5ca9deb55597a115a6de182ca2aef3c4497bd7d6) as to how I'm importing it.
rg with 'llama-swap' should find any relevant bits, if you'd like to help me out.

06kellyjac commented:
Ah. `llama-swap-149-go-modules.drv`. This is probably because I had a cached copy of the go modules beforehand.

anund commented:
Looks like #417516 was working on this but stalled at version 137. (connecting the issues)

06kellyjac commented:
Ah, thanks for the link.

anund commented:
Copied locally and swapped to version 150, works for me. :+1:

bachp commented:
@06kellyjac You might want to take a look at how the UI is handled in https://github.com/NixOS/nixpkgs/pull/417516. It builds it separately fromt he go module.

I think it is simpler as you can re-use `buildNpmPackage` and don't need to add npm machinery into `buildGoModule`.

@podium868909 I hope you are OK if we focus the effort for the package in this PR. Ideally we can also include https://github.com/NixOS/nixpkgs/pull/417516/commits/1fdfe7f6b886a8576c841c400302fb6694533002 here.

06kellyjac commented:
:+1: 
but I also want to see if `nix-update` can just handle it with the split out UI

06kellyjac committed:
nixos/llama-swap: add nixos vm test with opt-in model usage

Add a basic llama-swap test and perform more extensive tests using
SmolLM2 if allowed.

06kellyjac commented:
@bachp ok, I've updated and added separate UI derivation. I think the way I've done it is pretty clean and readable.
I've confirmed `nix-update` is fine with updating all the hashes.

I tweaked the module docs a bit & added the hardening from `llama-cpp`
I also added a test with opt-in model usage. See main post for details.
The full test works fine with the hardening.

I'm not 100% sure on if the `literalExpression` is accurate when rendered in the docs so I do want to double check that.

---

CC @podium868909 to confirm you're happy to be a maintainer of `llama-swap` before we merge
